component/main.jsonnet

```
@@ -81,80 +86,154 @@ local alerts = {
-  local rules = params.managedResources[mrName]._rules;
-  local role = roles.generateRole(mrName, rules, false);
-  local roleBinding = roles.generateBinding(mrName, role.metadata.name, false);
-  [ role, roleBinding ] else [];
+  };
+
+local clusterRole(prefix) =
+  function(path)
+    role(prefix, null)(path) + {
+      kind: 'ClusterRole',
+      metadata+: {
+        namespace:: null,
+      },
+    };
 
-local clusterRoleBinding(mrName) = [
-  roles.generateBinding(mrName, roleName, true)
-  for roleName in com.renderArray(std.get(params.managedResources[mrName], '_clusterRoles', []))
-];
+local roleBinding(roleNamePrefix) =
+  function(roleNs, roleName, saNs, saName) {
+    local bindingName = std.join(':', std.prune([ 'espejote', 'supplemental', roleName, if saNs != roleNs then saNs, saName ])),
+    apiVersion: 'rbac.authorization.k8s.io/v1',
+    kind: 'RoleBinding',
+    metadata: {
+      labels: {
+        'app.kubernetes.io/name': bindingName,
+      },
+      name: bindingName,
+      [if roleNs != null then 'namespace']: roleNs,
+    },
+    roleRef: {
+      apiGroup: 'rbac.authorization.k8s.io',
+      kind: 'Role',
+      name: roleNamePrefix + roleName,
+    },
+    subjects: [
+      {
+        kind: 'ServiceAccount',
+        name: saName,
+        namespace: saNs,
+      },
+    ],
+  };
+
+local clusterRoleBinding(roleNamePrefix) =
+  function(roleName, saNs, saName)
+    roleBinding(roleNamePrefix)(null, roleName, saNs, saName) + {
+      kind: 'ClusterRoleBinding',
+    };
+
+local roleBindingsForManagedResourceAndRoles(roleNamePrefix) =
+  function(managedResourcePath, rolePaths)
+    local mrNs = namespacedName(managedResourcePath).namespace;
+    std.map(function(rp)
+      local roleNsName = namespacedName(rp, namespace=mrNs);
+      roleBinding(roleNamePrefix)(
+        roleNsName.namespace,
+        roleNsName.name,
+        mrNs,
+        serviceAccountName(managedResourcePath),
+      ), rolePaths);
+
+local clusterRoleBindingsForManagedResourceAndRoles(roleNamePrefix) =
+  function(managedResourcePath, rolePaths)
+    std.map(function(rp) clusterRoleBinding(roleNamePrefix)(
+      namespacedName(rp).name,
+      namespacedName(managedResourcePath).namespace,
+      serviceAccountName(managedResourcePath),
+    ), rolePaths);
+
+local supplementalRoles = {
+  ['43_supplemental_role_%(namespace)s_%(name)s' % namespacedName(path)]:
+    local roles = std.get(params.managedResources[path], '_roles', {});
+    local mrNsName = namespacedName(path);
+    local roleNamePrefix = std.join(':', [ 'espejote', 'supplemental', mrNsName.namespace, mrNsName.name, '' ]);
+    com.generateResources(roles, role(roleNamePrefix, mrNsName.namespace)) +
+    roleBindingsForManagedResourceAndRoles(roleNamePrefix)(path, std.objectFields(roles)) +
+    roleBindingsForManagedResourceAndRoles(roleNamePrefix)(path, std.get(params.managedResources[path], '_roleBindings', []))
+  for path in std.objectFields(params.managedResources)
+};
```

maybe prune the resulting object to avoid emitting empty files?

```suggestion
local supplementalRoles = std.prune({
  ['43_supplemental_role_%(namespace)s_%(name)s' % namespacedName(path)]:
    local roles = std.get(params.managedResources[path], '_roles', {});
    local mrNsName = namespacedName(path);
    local roleNamePrefix = std.join(':', [ 'espejote', 'supplemental', mrNsName.namespace, mrNsName.name, '' ]);
    com.generateResources(roles, role(roleNamePrefix, mrNsName.namespace)) +
    roleBindingsForManagedResourceAndRoles(roleNamePrefix)(path, std.objectFields(roles)) +
    roleBindingsForManagedResourceAndRoles(roleNamePrefix)(path, std.get(params.managedResources[path], '_roleBindings', []))
  for path in std.objectFields(params.managedResources)
});
```